### PR TITLE
[SYNPY-1344] Adding activity model for OOP

### PR DESCRIPTION
**Pull request by BryanFauble**
**Problem:**

1. We need to create an OOP approach to working with the Activity model.
2. Some documentation needed to be updated
3. Remove OTEL from the test scripts showing how the interfaces are working

**Solution:**

1. Creating the Activity model based off the Rest API
2. Updating documentation
3. Creating a test script

**Testing:**

1. Unit and integration tests
2. The `oop_poc_activity.py` script

### CODE REVIEW

File: docs/scripts/object_orientated_programming_poc/oop_poc_activity.py
Comment (thomasyu888) [inline docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:53]: Does the replace the old `Activity` object or do we have two `Activity` objects now?

Comment (BryanFauble) [reply docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:53]: For now we have 2 Activity objects. I intend to deprecate all of the current models (Entity, File, Folder, Activity ect....) when we have everything ready.

File: docs/scripts/object_orientated_programming_poc/oop_poc_activity.py
Comment (BWMac) [inline docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:23]: Can you update the sharing settings on your example project so that at least DPE team members can run this script?

Comment (BryanFauble) [reply docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:23]: I added the DPE team to this project

File: synapseclient/models/activity.py
Comment (jaymedina) [inline synapseclient/models/activity.py:50]: nit: name

File: synapseclient/models/activity.py
Comment (jaymedina) [inline synapseclient/models/activity.py:42]: Maybe specify what this URL should be pointing to

**Review by jaymedina — COMMENTED** [review]
Looks good at first glance, I made some small comments, most are just for clarity, feel free to ignore some of them.

File: docs/scripts/object_orientated_programming_poc/oop_poc_activity.py
Comment (jaymedina) [reply docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:100]: You address these things in the MD file so I'll resolve this.

File: docs/scripts/object_orientated_programming_poc/oop_poc_project.py
Comment (BryanFauble) [reply docs/scripts/object_orientated_programming_poc/oop_poc_project.py:64]: Added

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py:42]: Adding this from the synapse help docs: "such as a link to a GitHub commit or a link to a specific version of a software tool."

Comment (BryanFauble) [reply synapseclient/models/activity.py:42]: https://help.synapse.org/docs/Provenance.1972470373.html#Provenance-OverviewofSynapseProvenance

File: docs/scripts/object_orientated_programming_poc/oop_poc_activity.py
Comment (BryanFauble) [reply docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:96]: Added:
```
    # These 3 will be the same activity
    print("Modified activity showing name is updated for all files:")
    print((await new_activity_instance.store()).name)
    print((await Activity.from_parent(parent=file)).name)
    print((await Activity.from_parent(parent=second_file)).name)
```

File: docs/scripts/object_orientated_programming_poc/oop_poc_activity.py
Comment (thomasyu888) [inline docs/scripts/object_orientated_programming_poc/oop_poc_activity.py:1]: I'll leave it to the team to validate these in the feature branch.  I will be running through these after they are merged into develop / client release, but awesome work here!

File: synapseclient/core/utils.py
Comment (thomasyu888) [inline synapseclient/core/utils.py:1236]: Nice, do we want to include the `opentelemetry_context` parameter here to let people know about concurrent operations?

File: synapseclient/models/activity.py
Comment (thomasyu888) [inline synapseclient/models/activity.py:23]: Nit: Semantics, but this is `UsedEntityReference`: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Reference.html.  `UsedEntity`: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/provenance/UsedEntity.html


Did we want this class to be consistent with the API?  I know there is logic that relies on this class being this way.

File: synapseclient/models/activity.py
Comment (thomasyu888) [inline synapseclient/models/activity.py:40]: Nit: UsedURL is: https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/provenance/UsedURL.html

File: synapseclient/models/activity.py
Comment (thomasyu888) [inline synapseclient/models/activity.py:116]: I'm a little confused with the `synapse_activity` type and the description.  If its response from API, would it be `dict`?  I see that you're also supporting the return object of `syn.getActivity`

File: synapseclient/models/activity.py
Comment (thomasyu888) [inline synapseclient/models/activity.py:207]: Nit: not to make things more complex, but tuples are irritating at times when keeping order.  What are your thoughts on `namedtuples`?  (Note: I'm good with this, just opening a discussion.)

File: synapseclient/models/file.py
Comment (thomasyu888) [inline synapseclient/models/file.py:114]: Wonderful to see the composition come together!

**Review by thomasyu888 — APPROVED** [review]
🔥 Wonderful PR. I left some comments, but I'm going to pre-approve.  Most of them are Nits and/or potentially different Jira tickets.

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py:23]: I had thought about having 2 layers for this to model exactly how the REST API represents there. However, I elected to follow what I did in the code here to remove the nested objects.

I also removed `wasExecuted` and `concreteType` anywhere from the OOP model attributes. Instead separating them out into 2 different lists on the Activity. IMO this makes it a bit easier to work with by having separate collections of items rather than a boolean that controls it.

```
    used: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
    """The entities used by this Activity."""

    executed: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
    """The entities executed by this Activity."""
```

Thoughts on how I am diverging from the REST API here?

One of the possible future downsides here is that if the synapse REST API changes in such a way that conflicts with my implementation, I'd need to go back on some of these decisions.

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py:116]: In the `client.py` the save action calls down to this method:
```
    def _saveActivity(self, activity: Activity) -> Activity:
        """
        Save the Activity

        Arguments:
            activity: The Activity to be saved

        Returns:
            An Activity object
        """
        if "id" in activity:
            # We're updating provenance
            uri = "/activity/%s" % activity["id"]
            activity = Activity(data=self.restPUT(uri, json.dumps(activity)))
        else:
            activity = self.restPOST("/activity", body=json.dumps(activity))
        return activity
```


Specifically the return type is `Activity`: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/SYNPY-1344-activity-model/synapseclient/activity.py#L133

I want to get away from having this 

1. `from synapseclient.models.activity import Activity` and
2. `from synapseclient.activity import Activity as Synapse_Activity`

So we don't have 2 models. In the end though I need an exact object of what the REST API is expecting/returning, and what the OOP model looks like; A translation between the two. I am hoping do something with this in https://sagebionetworks.jira.com/browse/SYNPY-1411 - Or perhaps when the open API generated client drops I am use that and automatically get "what the REST API is expecting/returning", and write keep the translation between the 2 in the current python client (This translation is what the `fill_from_dict` and [this line is doing](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/SYNPY-1344-activity-model/synapseclient/models/activity.py#L239))

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py]: I like the idea of namedtuples - I added this in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1055/commits/a38f2c2bfa771afdd0dad03a9382bce311b98ce4

File: synapseclient/core/utils.py
Comment (BryanFauble) [reply synapseclient/core/utils.py]: Added more context in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1055/commits/a38f2c2bfa771afdd0dad03a9382bce311b98ce4

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py]: We do! I added it to https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1055/commits/a38f2c2bfa771afdd0dad03a9382bce311b98ce4

File: synapseclient/models/activity.py
Comment (thomasyu888) [reply synapseclient/models/activity.py:131]: Got it, I think it was confusing from the function name since `fill_from_dict` seems like it will take in a `dict`.  Although... the function does look like it would support a `dict. ha.

File: synapseclient/models/activity.py
Comment (thomasyu888) [reply synapseclient/models/activity.py:25]: This makes sense to me.  Thanks!

File: synapseclient/models/activity.py
Comment (BryanFauble) [reply synapseclient/models/activity.py:131]: I suppose I could add a union type hint here too

File: synapseclient/models/activity.py
Comment (thomasyu888) [reply synapseclient/models/activity.py:42]: Discussed in other comments on why this is the way it is.

**Pull request merged**
